### PR TITLE
PLAT-648 Delete Old Program Executions Automatically

### PR DESCRIPTION
**Pull request by thees-k**
Summary:
* Add new column "executionRetentionDays" to `AGProgram` and `AGProgramLog`
* Add `ProgramExecutionsDeleter` and `ProgramExecutionsDeleterTest`
* Add new program `ProgramExecutionsCleanupProgram`

Test Plan:
* See `ProgramExecutionsDeleterTest`
* See Programs Page:

* See Program Executions Page:

### CODE REVIEW

File: platform-core-module/src/main/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionDeleter.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+		Log.finfo("%s started for %s", ProgramExecutionDeleter.class.getSimpleName(), referenceDay);
+
+		for (AGProgram program: AGProgram.TABLE.createSelect().orderBy(AGProgram.ID)) {
+			deleteExecutionsOfProgram(program);
+		}
+	}
+
+	private void deleteExecutionsOfProgram(AGProgram program) {
+
+		Log.finfo("Cleanup executions of program %s", fetchProgramNameIfPossible(program));
```
Comment (alexander-schmidt) [inline platform-core-module/src/main/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionDeleter.java:45]: `cleanup` is not a verb... this should be reworded.

File: platform-core-module/src/main/java/com/softicar/platform/core/module/CoreI18n.java [hunk 2]
```diff
@@ -767,6 +767,10 @@
 		.de("Reply-To");
 	I18n0 RESET_PASSWORD = new I18n0("Reset Password")//
 		.de("Passwort zurücksetzen");
+	I18n0 RETENTION_DAYS_OF_EXECUTIONS = new I18n0("Retention Days of Executions")//
+		.de("Speicherungstage für Ausführungen");
+	I18n0 RETENTION_DAYS_OF_EXECUTIONS_MUST_NOT_BE_SMALLER_THAN_0 = new I18n0("Retention days of executions must not be smaller than 0.")//
```
Comment (alexander-schmidt) [inline platform-core-module/src/main/java/com/softicar/platform/core/module/CoreI18n.java:772]: I suggest to reword this, in order to get rid of the negation, e.g.:
`Retention days of executions must be at least 0.`
It's easier to grasp this way.

File: platform-core-module/src/main/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionDeleter.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+import com.softicar.platform.common.date.Day;
+import com.softicar.platform.common.date.DayTime;
+import com.softicar.platform.core.module.program.AGProgram;
+import com.softicar.platform.core.module.program.execution.AGProgramExecution;
+import com.softicar.platform.core.module.program.execution.AGProgramExecutionLog;
+import com.softicar.platform.db.core.transaction.DbTransaction;
+import java.util.List;
+
+/**
+ * Deletes program execution records ({@link AGProgramExecution}) and their log
```
Comment (alexander-schmidt) [inline platform-core-module/src/main/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionDeleter.java:16]: - We could remove some redundancy here, like: `Deletes ({@link AGProgramExecution}) records and corresponding ({@link AGProgramExecutionLog}) records [...]`
- Also, there's `[...] records records [...]` in the Javadoc.

File: platform-core-module/src/main/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionDeleter.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+ * records records ({@link AGProgramExecutionLog}) if they are older than the
+ * return value of related method
+ * {@link com.softicar.platform.core.module.program.AGProgram#getRetentionDaysOfExecutions()}
+ *
+ * @author Thees Koester
+ */
+public class ProgramExecutionDeleter {
```
Comment (alexander-schmidt) [inline platform-core-module/src/main/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionDeleter.java:23]: Could be renamed to `ProgramExecutionsDeleter` (mind the plural), to indicate that it commonly deletes several execution records (not: just one).

File: platform-core-module/src/main/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionDeleter.java [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+			.getFirstAsOptional()
+			.map(AGProgramExecution::getTerminatedAt)
+			.map(DayTime::getDay)
+			.ifPresentOrElse(minimalDayInDb -> deleteBatchwise(program, minimalDayInDb), this::logInfoNothingToDelete);
+	}
+
+	private String fetchProgramNameIfPossible(AGProgram program) {
+
+		try {
+			// Will fail in JUnit tests
```
Comment (alexander-schmidt) [inline platform-core-module/src/main/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionDeleter.java:61]: - What about this comment? Why is that so? What can we do about it?
- The catch block below looks super scary; should not be necessary here.

File: platform-core-module/src/test/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionDeleterTest.java [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+	}
+
+	private void insertLogRecord(AGProgramExecution programExecution) {
+
+		AGTransaction transaction = new AGTransaction().setAt(DayTime.now()).setByToCurrentUser().save();
+		AGProgramExecutionLog.TABLE.getOrCreate(new Tuple2<>(programExecution, transaction));
+	}
+
+	@Test
+	public void testDelete() {
```
Comment (alexander-schmidt) [inline platform-core-module/src/test/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionDeleterTest.java:101]: - In pretty much all of our tests, the `@Test` methods are situated directly below the constructor. We should move this method accordingly.
- I think we need further test methods for edge cases, e.g.:
  - empty executions table
  - nonempty executions table but nothing to do for any program
  - ... things like that.

File: platform-core-module/src/main/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionDeleter.java [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+
+		List<AGProgramExecution> programExecutions = loadToBeDeletedProgramExecutions(program, thresholdDay);
+		Log.finfo("Deleting %d program execution(s) older than %s...", programExecutions.size(), thresholdDay);
+		try (DbTransaction transaction = new DbTransaction()) {
+			deleteProgramExecutionLogsAndProgramExecutions(programExecutions);
+			transaction.commit();
+		}
+	}
+
+	private List<AGProgramExecution> loadToBeDeletedProgramExecutions(AGProgram program, Day thresholdDay) {
```
Comment (alexander-schmidt) [inline platform-core-module/src/main/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionDeleter.java:97]: Not sure if `zu löschende` can be translated this way. How about `loadOutdatedProgramExecutions`?

File: platform-core-module/src/main/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionDeleter.java [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+
+		return AGProgramExecution.TABLE
+			.createSelect()
+			.where(AGProgramExecution.PROGRAM_UUID.equal(program.getProgramUuid()))
+			.where(AGProgramExecution.TERMINATED_AT.isNotNull())
```
Comment (alexander-schmidt) [inline platform-core-module/src/main/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionDeleter.java:102]: I think this line can be omitted because, in SQL, a comparison with `NULL`
- does not cause an exception, and
- always yields `false` (so that records with `TERMINATED_AT = NULL` will never be in the result set).

File: platform-core-module/src/main/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionCleanupProgram.java [hunk 9]
```diff
@@ -0,0 +1,26 @@
+package com.softicar.platform.core.module.program.execution.cleanup;
+
+import com.softicar.platform.core.module.program.IProgram;
+import com.softicar.platform.core.module.program.execution.AGProgramExecution;
+import com.softicar.platform.core.module.program.execution.AGProgramExecutionLog;
+import com.softicar.platform.emf.source.code.reference.point.EmfSourceCodeReferencePointUuid;
+
+/**
+ * Deletes program execution records ({@link AGProgramExecution}) and their log
```
Comment (alexander-schmidt) [inline platform-core-module/src/main/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionCleanupProgram.java:9]: - We could remove some redundancy here, like: `Deletes ({@link AGProgramExecution}) records and corresponding ({@link AGProgramExecutionLog}) records [...]`
- Also, there's `[...] records records [...]` in the Javadoc.

File: platform-core-module/src/test/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionDeleterTest.java [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+
+		new ProgramExecutionDeleter(0).delete();
+		assertEquals(expectedRetainedProgramExecutionsMap, loadProgramExecutions());
```
Comment (alexander-schmidt) [inline platform-core-module/src/test/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionDeleterTest.java:104]: Assertions should be more explicit.
To me, the content of `expectedRetainedProgramExecutionsMap` at this point is totally unclear... I need to read and understand the whole test setup to comprehend the expectation in this assertion.
I think this needs to be simplified, and/or made more explicit - e.g. by explicitly defining the expected executions in this test method.

File: platform-emf/src/main/resources/com/softicar/platform/emf/emf-form-style.css [hunk 11]
```diff
@@ -98,11 +98,6 @@
 	box-shadow: var(--DOM_BUTTON_DECORATION_INTERACTION_BOX_SHADOW);
 }
 
-.EmfFormPrimaryActionSectionDiv .EmfFormSectionContentDiv:not(.invisible) .DomButton > .DomButtonIcon {
```
Comment (alexander-schmidt) [inline platform-emf/src/main/resources/com/softicar/platform/emf/emf-form-style.css:101]: This PR is completely unrelated to CSS. This change should not be here.

**Review by alexander-schmidt — CHANGES_REQUESTED** [review]

File: platform-core-module/src/main/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionDeleter.java (hunk 1 above)
Comment (thees-k) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionDeleter.java:45]: done

File: platform-core-module/src/main/java/com/softicar/platform/core/module/CoreI18n.java (hunk 2 above)
Comment (thees-k) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/CoreI18n.java:772]: done

File: platform-core-module/src/main/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionDeleter.java (hunk 3 above)
Comment (thees-k) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionDeleter.java:16]: done

File: platform-core-module/src/main/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionDeleter.java (hunk 4 above)
Comment (thees-k) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionDeleter.java:23]: Done.
I also renamed:
* `ProgramExecutionCleanupProgram` -> `ProgramExecutionsCleanupProgram`
* `ProgramExecutionDeleterTest` -> `ProgramExecutionsDeleterTest`

File: platform-core-module/src/main/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionDeleter.java (hunk 7 above)
Comment (thees-k) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionDeleter.java:97]: I renamed the method to `loadProgramExecutionsThatAreToBeDeleted`

File: platform-core-module/src/main/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionDeleter.java (hunk 8 above)
Comment (thees-k) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionDeleter.java:102]: done

File: platform-core-module/src/main/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionCleanupProgram.java (hunk 9 above)
Comment (thees-k) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionCleanupProgram.java:9]: done

File: platform-emf/src/main/resources/com/softicar/platform/emf/emf-form-style.css (hunk 11 above)
Comment (thees-k) [reply platform-emf/src/main/resources/com/softicar/platform/emf/emf-form-style.css:101]: I really should have seen that!
Done.

File: platform-core-module/src/test/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionDeleterTest.java (hunk 10 above)
Comment (thees-k) [reply platform-core-module/src/test/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionDeleterTest.java]: I improved the test: It should be easier to understand now. I added more test methods to validate different scenarios.
Please have a look.

File: platform-core-module/src/test/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionDeleterTest.java (hunk 6 above)
Comment (thees-k) [reply platform-core-module/src/test/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionDeleterTest.java]: I improved `ProgramExecutionDeleterTest`. Please have a look.

File: platform-core-module/src/main/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionDeleter.java (hunk 5 above)
Comment (thees-k) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionDeleter.java]: I wasn't happy with the method either.
I tried to mock real programs now. Please have a look.

File: platform-core-module/src/test/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionsDeleterTest.java [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+	}
+
+	private SetMap<AGProgram, AGProgramExecution> loadProgramExecutions() {
+
+		SetMap<AGProgram, AGProgramExecution> programExecutions = new SetMap<>();
+		AGProgramExecution.createSelect().forEach(record -> programExecutions.addToSet(AGProgram.loadOrInsert(record.getProgramUuid()), record));
+		return programExecutions;
+	}
+
+	private void insertProgramWithZeroRetentionDaysOfExecutionsAndInsertItsExecutions() {
```
Comment (alexander-schmidt) [inline platform-core-module/src/test/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionsDeleterTest.java:116]: Those `insertProgramWith*` methods basically perform the test setup _and_ define expectations for the assertions. We usually separate those aspects/concerns.
This _can_ surely be done this way. I just find it difficult to wrap my head around this.

**Review by alexander-schmidt — CHANGES_REQUESTED** [review]

File: platform-core-module/src/main/java/com/softicar/platform/core/module/CoreI18n.java [hunk 13]
```diff
@@ -767,6 +767,10 @@
 		.de("Reply-To");
 	I18n0 RESET_PASSWORD = new I18n0("Reset Password")//
 		.de("Passwort zurücksetzen");
+	I18n0 RETENTION_DAYS_OF_EXECUTIONS = new I18n0("Retention Days of Executions")//
```
Comment (alexander-schmidt) [inline platform-core-module/src/main/java/com/softicar/platform/core/module/CoreI18n.java:770]: We should probably simplify the English label (and column name): `Execution Retention Days`... is easier to grasp. German translation could stay as-is.

File: platform-core-module/src/test/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionsDeleterTest.java [hunk 14]
```diff
@@ -0,0 +1,238 @@
...
+		return programExecution;
+	}
+
+	private void insertLogRecord(AGProgramExecution programExecution) {
+
+		AGTransaction transaction = new AGTransaction().setAt(DayTime.now()).setByToCurrentUser().save();
+		AGProgramExecutionLog.TABLE.getOrCreate(new Tuple2<>(programExecution, transaction));
+	}
+
+	private enum ExecutionDay {
```
[the remark this answers is not in the document]
Comment (thees-k) [reply platform-core-module/src/test/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionsDeleterTest.java:198]: done

File: platform-core-module/src/test/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionsDeleterTest.java (hunk 12 above)
Comment (thees-k) [reply platform-core-module/src/test/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionsDeleterTest.java:116]: well, I think, doing setup and defining expected values at the same time saves us many code lines

File: platform-core-module/src/test/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionsDeleterTest.java [hunk 15]
```diff
@@ -0,0 +1,238 @@
...
+		@Override
+		public void executeProgram() {
+
+			// Nothing to do
+		}
+	}
+
+	@TestingOnly
+	@EmfSourceCodeReferencePointUuid("c977d6ed-ad3d-4748-a7fc-1a64be5c4728")
+	public static class TestProgram1 extends AbstractTestProgram {
```
[the remark this answers is not in the document]
Comment (thees-k) [reply platform-core-module/src/test/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionsDeleterTest.java:223]: Unfortunately not, I tested it. It has to be public.
Maybe the `@TestingOnly` above the classes will help in future static code checks.

File: platform-core-module/src/main/java/com/softicar/platform/core/module/CoreI18n.java (hunk 13 above)
Comment (thees-k) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/CoreI18n.java:770]: done

File: platform-core-module/src/test/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionsDeleterTest.java (hunk 12 above)
Comment (alexander-schmidt) [reply platform-core-module/src/test/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionsDeleterTest.java]: I think unit tests are all about clarity and explicitness, and not so much about low redundancy. In that regard, they're quite different from regular code.
My point is: There's only one single `assertEquals` method in that test class. It compares two maps.
If this assertion ever fails, the cause is _not_ obvious. All we know in this case is, that an "assertion failed" (the equals-based comparison of two maps). We would then need to start _debugging the unit test_ (!) in order to identify the actual problem.

Comment (thees-k) [reply platform-core-module/src/test/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionsDeleterTest.java]: Okay, I changed the test class.

**Review by alexander-schmidt — APPROVED** [review]
